src/Evaluator.hs

```diff
@@ -278,22 +278,21 @@ withScopeOf scope expr = do
 
 doDataCase :: SrcLoc -> Value -> [(ConstructorPattern, Core)] -> Eval Value
 doDataCase loc v0 [] = throwError (EvalErrorCase loc v0)
-doDataCase loc v0 ((pat, rhs) : ps) = match (doDataCase loc v0 ps) pat
+doDataCase loc v0 ((pat, rhs) : ps) =
+  match (doDataCase loc v0 ps) (eval rhs) [(view constructorPattern pat, v0)]
   where
```

Comment:
I think a type signature with comments would make this a lot more comprehensible:

```haskell
match
  :: Eval Value  -- ^ failure continuation
  -> Eval Value  -- ^ success continuation, to be ran in an extended environment
  -> [(ConstructorPatternF ConstructorPattern, Core)]  -- ^ [(pattern, scrutinee)]
  -> Eval Value
```

Reply:
Done!